### PR TITLE
Another referential transparent API refactoring

### PR DESCRIPTION
**Pull request by mariusmuja**
I refactored the referential transparent API proposed in #68 so that it keeps the client code mostly unchanged (similarly to what the current master API does) but without wrapping the IO monad.

### CODE REVIEW

**Comment by fdietze** [comment]
Thank you for working on this! I'll have a detailed look at it as soon as possible.

**Comment by mariusmuja** [comment]
@fdietze As far as I can tell it is referential transparent (what makes you think it isn't ?), for example for any `makeComponent()`:
```scala
val component = makeComponent()

val comp1 = div(component, component) 
val comp2 = div(makeComponent(), makeComponent())
```
`comp1` and `comp2` are the same.

I just added a test for this (96fc5fe), which seems to fail both for the master implementation and the one in #68, because the issue fixed by a448d1b.

File: src/test/scala/outwatch/ScenarioTestSpec.scala [hunk 1]
```diff
@@ -91,6 +91,38 @@ class ScenarioTestSpec extends UnitSpec with BeforeAndAfterEach {
     document.getElementById("greeting").innerHTML shouldBe greetStart + name2
   }
 
+  "A component" should "be referential transparent" in {
```
Comment (fdietze) [inline src/test/scala/outwatch/ScenarioTestSpec.scala]: :+1:
This test is a very important contribution!

**Comment by fdietze** [comment]
@mariusmuja Thanks for clarifying.

I see, as everything happens in IO, asProxy can never be called outside and is therefore safe.

**Comment by fdietze** [comment]
I'm actually in favor of this PR now.

What is left, is a convenient way to write inline-styles.

currently one needs to write:
```scala
IO.pure(Style("color", "red"))
```

There should be a simple solution until `scala-dom-types` is integrated. Maybe just provide a function `outwatch.dom.style` like found in the tests of this PR?:
```scala
def style(property: String, value: String): IO[Style] = IO.pure(Style(property: String, value: String))
```

**Comment by fdietze** [comment]
I just had the idea to drive this approach of flattening out IO even further:
```scala
  def -->(sink: Sink[E]):IO[EventEmitter[E]] =
    IO.pure(EventEmitter(eventType, sink.observer))

  def -->(sinkIO: IO[Sink[E]]):IO[EventEmitter[E]] =
    sinkIO.flatMap(-->)
```

In some cases this makes it possible to write components without for-expressions. The downside is that it looks like the old api, but unintuitively fails when combining streams. In many cases a fallback to the for-loop is needed. Any ideas?

I implemented the methods on this branch: https://github.com/mariusmuja/outwatch/compare/referential-transparent-refactor...OutWatch:referential-transparent-refactor-flatten-stream-io

**Comment by mariusmuja** [comment]
@fdietze Yeah, I though about doing that a while ago (https://github.com/OutWatch/outwatch/issues/40#issuecomment-312546861), but it doesn't work as expected:

For example in this code:
```scala
val handler = createStringHandler()
div(
  button(click("clicked") --> handler),
  div(`class` := "label", child <-- handler)
)
```
the `div` doesn't get updated on `button` click, because `createStringHandler()` is now referential transparent, so it's equivalent to writing:
```scala
div(
  button(click("clicked") --> createStringHandler()),
  div(`class` := "label", child <-- createStringHandler())
)
```

**Comment by fdietze** [comment]
I noticed that Sink.create() does not return an IO. Is this on purpose?

**Comment by fdietze** [comment]
Here is a useful diff for reviewing: https://github.com/OutWatch/outwatch/compare/07a0f2d2d129bd9d1990f94140c90d4e73093d8a...mariusmuja:referential-transparent-refactor

**Comment by fdietze** [comment]
And should `Sink` and `Observable` be imported somehow when writing `import outwatch.dom._` ? I think they should. Because else, extra imports are needed when writing component-signatures which take inputs and outputs:
```scala
def comp(input:Sink[Int], output:Observable[Int]) = ...
```

File: src/main/scala/outwatch/dom/helpers/DomUtils.scala [hunk 2]
```diff
@@ -202,11 +202,9 @@ object DomUtils {
   }
 
   def render(element: Element, vNode: VNode): IO[Unit] = for {
```
[the remark this answers is not in the document]
Comment (fdietze) [reply src/main/scala/outwatch/dom/helpers/DomUtils.scala]: Note that:

```scala
type VNode = IO[VNode_]
```

You just got confused by the hidden IO-Monad ;)

[the remark this answers is not in the document]
Comment (fdietze) [reply src/main/scala/outwatch/dom/helpers/DomUtils.scala]: Maybe we should not hide it and sacrifice the type-alias. What do you think?

**Comment by mariusmuja** [comment]
I like the idea of having Observable and Sink imported by `outwatch.dom._`. We could do something like this:
```diff
 package object dom extends Attributes with Tags with Handlers {
 
+  type Observable[+A] = rxscalajs.Observable[A]
+  val Observable = rxscalajs.Observable
+
+  type Sink[-A] = outwatch.Sink[A]
+  val Sink = outwatch.Sink
+
   type VNode = IO[VNode_]
   type VDomModifier = IO[VDomModifier_]

```

An additional benefit is that it would ease porting Outwatch to another reactive streams library such as Monix. I prefer Monix to rxjs, I did a port here: https://github.com/mariusmuja/outwatch/tree/referential-transparent-refactor-monix

File: src/main/scala/outwatch/dom/package.scala [hunk 3]
```diff
@@ -14,9 +14,9 @@ package object dom extends Attributes with Tags with Handlers {
 
   implicit def optionIsEmptyModifier(opt: Option[VDomModifier]): VDomModifier = opt getOrElse IO.pure(EmptyVDomModifier)
 
-  implicit class ioVTreeMerge(tree: IO[VTree]) {
-    def apply(args: VDomModifier*): IO[VTree] = {
-      tree.flatMap(vtree => IO.pure(VTree(vtree.nodeType, vtree.modifiers ++ args)))
+  implicit class ioVTreeMerge(vnode: VNode) {
+    def apply(args: VDomModifier*): VNode = {
+      vnode.flatMap(vnode_ => vnode_(args:_*))
```
Comment (mariusmuja) [inline src/main/scala/outwatch/dom/package.scala]: :+1:

File: src/test/scala/outwatch/ScenarioTestSpec.scala [hunk 4]
```diff
@@ -153,18 +177,16 @@ class ScenarioTestSpec extends UnitSpec with BeforeAndAfterEach {
       state = adds.merge(deletes)
         .scan(Vector[String]())((state, modify) => modify(state))
         .map(_.map(n => TodoComponent(n, deleteHandler)))
-
-
-      root <- div(
-        TextFieldComponent("Todo: ", inputHandler),
+      textFieldComponent <- TextFieldComponent("Todo: ", inputHandler)
+    } yield div(
```
Comment (LukaJCB) [inline src/test/scala/outwatch/ScenarioTestSpec.scala]: Correct me, if I'm wrong, but here root is a `IO[IO[VNode]]` correct?

**Comment by LukaJCB** [comment]
I'd just like to say thank you to all of you who put in such great work! I'm really glad to see the discussion we're having and the progress we're making!

File: src/test/scala/outwatch/ScenarioTestSpec.scala (hunk 4 above)
Comment (mariusmuja) [reply src/test/scala/outwatch/ScenarioTestSpec.scala]: That is correct, it would end up flattened by the ``` vtree.flatMap``` below.

**Comment by LukaJCB** [comment]
Awesome, LGTM! I'd like to bikeshed about some of the naming of internal things, but that shouldn't affect the API, so I'm going to merge this. Thank you again @mariusmuja and @fdietze  and @cornerman! :heart:

**Comment by mariusmuja** [comment]
Yes, we should have better names for: `VNode_` and `VDomModifier_`.

**Pull request merged**